File: fetch_player_stats.py
#!/usr/bin/env python3
"""
Python script to fetch NFL player statistics by scraping ESPN game log pages.

This script is called from Node.js to get player stats.
ESPN's API doesn't provide direct player stats endpoints, so we scrape
their game log pages which display comprehensive player statistics.

Strategy:
1. Scrape ESPN player game log page (format: /nfl/player/gamelog/_/id/{id}/{slug})
2. Extract season totals from the "Regular Season Stats" row
3. Extract last game stats from the first game row
4. Return structured JSON data
"""

import sys
import json
import requests
import warnings
from bs4 import BeautifulSoup

# Suppress urllib3 SSL warnings
warnings.filterwarnings('ignore', category=UserWarning, module='urllib3')

def scrape_espn_gamelog(player_id, player_name=None):
    """
    Scrape ESPN player game log page to extract season statistics.
    
    ESPN URL format: https://www.espn.com/nfl/player/gamelog/_/id/{playerId}/{slug}
    If player_name is provided, we construct the slug from it.
    
    Args:
        player_id: ESPN player ID
        player_name: Optional player name to construct proper URL slug
        
    Returns:
        dict: Player statistics in JSON format, or None if error
    """
    # Construct slug from player name if provided
    if player_name:
        slug = player_name.lower().replace(" ", "-").replace(".", "").replace("'", "")
        url = f"https://www.espn.com/nfl/player/gamelog/_/id/{player_id}/{slug}"
    else:
        # Try without slug (may not work for all players)
        url = f"https://www.espn.com/nfl/player/gamelog/_/id/{player_id}"
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    
    try:
        response = requests.get(url, headers=headers, timeout=15, allow_redirects=True)
        
        if response.status_code != 200:
            return None
        
        soup = BeautifulSoup(response.content, 'html.parser')
        
        # Extract player name from page title if not provided
        extracted_name = player_name
        if not extracted_name:
            title_tag = soup.find('title')
            if title_tag:
                title_text = title_tag.get_text()
                # Title format: "Player Name Game Log - NFL" or similar
                match = title_text.split(' Game Log')[0] if ' Game Log' in title_text else None
                if match:
                    extracted_name = match.strip()
        
        # Find the game log table
        tables = soup.find_all('table')
        if not tables:
            return None
        
        first_table = tables[0]
        all_rows = first_table.find_all('tr')
        
        # Find the totals row (last row with "Regular Season Stats")
        totals_row = None
        for row in all_rows:
            cells = row.find_all(['td', 'th'])
            row_text = [cell.get_text(strip=True) for cell in cells]
            if len(row_text) > 0 and row_text[0] == 'Regular Season Stats':
                totals_row = row_text
                break
        
        # Parse game rows
        tbody = first_table.find('tbody')
        games = []
        
        if tbody:
            for row in tbody.find_all('tr'):
                cells = row.find_all(['td', 'th'])
                if len(cells) >= 10:
                    row_data = [cell.get_text(strip=True) for cell in cells]
                    # Skip totals row (starts with "Regular Season Stats")
                    if len(row_data) > 0 and row_data[0] == 'Regular Season Stats':
                        continue
                    try:
                        game = {
                            'date': row_data[0] if len(row_data) > 0 else '',
                            'opponent': row_data[1] if len(row_data) > 1 else '',
                            'result': row_data[2] if len(row_data) > 2 else '',
                            'rushing': {
                                'carries': row_data[3] if len(row_data) > 3 else '0',
                                'yards': row_data[4] if len(row_data) > 4 else '0',
                                'avg': row_data[5] if len(row_data) > 5 else '0',
                                'td': row_data[6] if len(row_data) > 6 else '0'
                            },
                            'receiving': {
                                'receptions': row_data[8] if len(row_data) > 8 else '0',
                                'targets': row_data[9] if len(row_data) > 9 else '0',
                                'yards': row_data[10] if len(row_data) > 10 else '0',
                                'td': row_data[12] if len(row_data) > 12 else '0'
                            }
                        }
                        games.append(game)
                    except:
                        continue
        
        # Extract season totals from totals row (more accurate than summing)
        if totals_row and len(totals_row) >= 5:
            # Format: ['Regular Season Stats', '222', '1,025', '4.6', '10', ...]
            # Index:  CAR=1, YDS=2, AVG=3, TD=4, ...
            total_rush_car = int(totals_row[1].replace(',', '')) if totals_row[1].replace(',', '').isdigit() else 0
            total_rush_yds = int(totals_row[2].replace(',', '')) if totals_row[2].replace(',', '').isdigit() else 0
            total_rush_avg = float(totals_row[3]) if totals_row[3].replace('.', '').replace('-', '').isdigit() else 0
            total_rush_td = int(totals_row[4].replace(',', '')) if totals_row[4].replace(',', '').isdigit() else 0
        elif games:
            # Fallback: calculate from individual games
            total_rush_yds = sum(int(g['rushing']['yards']) for g in games if g['rushing']['yards'].isdigit())
            total_rush_td = sum(int(g['rushing']['td']) for g in games if g['rushing']['td'].isdigit())
            total_rush_car = sum(int(g['rushing']['carries']) for g in games if g['rushing']['carries'].isdigit())
            total_rush_avg = round(total_rush_yds / total_rush_car, 1) if total_rush_car > 0 else 0
        else:
            total_rush_car = total_rush_yds = total_rush_td = total_rush_avg = 0
        
        # Get most recent game (last in list - games are listed chronologically, oldest to newest)
        last_game = games[-1] if games else None
        
        # Build result JSON
        result = {
            "success": True,
            "playerName": extracted_name or "Unknown",
            "playerId": player_id,
            "seasonStats": {
                "rushing": {
                    "carries": total_rush_car,
                    "yards": total_rush_yds,
                    "touchdowns": total_rush_td,
                    "average": total_rush_avg
                }
            },
            "lastGame": {
                "date": last_game['date'],
                "opponent": last_game['opponent'],
                "result": last_game['result'],
                "rushing": {
                    "yards": int(last_game['rushing']['yards']) if last_game['rushing']['yards'].isdigit() else 0,
                    "touchdowns": int(last_game['rushing']['td']) if last_game['rushing']['td'].isdigit() else 0
                }
            } if last_game else None,
            "source": "espn_gamelog",
            "gamesPlayed": len(games)
        }
        
        return result
        
    except Exception as e:
        return None

def fetch_player_stats(player_id, player_name=None):
    """
    Main function to fetch player statistics.
    
    Args:
        player_id: ESPN player ID
        player_name: Optional player name to construct proper URL slug
        
    Returns:
        dict: Player statistics in JSON format
    """
    try:
        result = scrape_espn_gamelog(player_id, player_name)
        
        if result:
            return result
        else:
            return {
                "success": False,
                "error": f"Could not scrape ESPN game log page for player ID {player_id}",
                "playerId": player_id,
                "suggestion": "Player ID may be incorrect, or ESPN page structure has changed."
            }
            
    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "playerId": player_id
        }

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({
            "success": False,
            "error": "Player ID required as argument"
        }))
        sys.exit(1)
    
    player_id = sys.argv[1]
    player_name = sys.argv[2] if len(sys.argv) > 2 else None
    result = fetch_player_stats(player_id, player_name)
    
    # Output JSON so Node.js can parse it
    print(json.dumps(result))

File: fetch_player_stats.js
#!/usr/bin/env node
/**
 * Node.js script to fetch NFL player statistics by scraping ESPN game log pages.
 * 
 * This script is called from server.js to get player stats.
 * ESPN's API doesn't provide direct player stats endpoints, so we scrape
 * their game log pages which display comprehensive player statistics.
 * 
 * Usage: node fetch_player_stats.js <playerId> [playerName]
 * Example: node fetch_player_stats.js 3043078 "Derrick Henry"
 */

const https = require('https');
const http = require('http');

/**
 * Fetch HTML content from a URL
 * Uses Node.js built-in https/http modules (no external dependencies)
 */
function fetchUrl(url) {
  return new Promise((resolve, reject) => {
    const client = url.startsWith('https:') ? https : http;
    
    client.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
      }
    }, (res) => {
      if (res.statusCode !== 200) {
        reject(new Error(`HTTP ${res.statusCode}: ${res.statusMessage}`));
        return;
      }
      
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve(data));
    }).on('error', reject);
  });
}

/**
 * Simple HTML parser to extract table data
 * Since we can't use external libraries, we'll parse HTML with regex/string manipulation
 */
function parseGameLogTable(html) {
  const games = [];
  let totalsRow = null;
  
  // Find the game log table
  // ESPN uses <table> tags with class "Table"
  const tableMatch = html.match(/<table[^>]*class="[^"]*Table[^"]*"[^>]*>([\s\S]*?)<\/table>/);
  if (!tableMatch) {
    return { games: [], totalsRow: null };
  }
  
  const tableHtml = tableMatch[1];
  
  // Find all rows
  const rowRegex = /<tr[^>]*>([\s\S]*?)<\/tr>/g;
  let rowMatch;
  
  while ((rowMatch = rowRegex.exec(tableHtml)) !== null) {
    const rowHtml = rowMatch[1];
    
    // Extract cell text from this row
    const cellRegex = /<t[dh][^>]*>(.*?)<\/t[dh]>/g;
    const cells = [];
    let cellMatch;
    
    while ((cellMatch = cellRegex.exec(rowHtml)) !== null) {
      // Remove HTML tags and decode entities
      let cellText = cellMatch[1]
        .replace(/<[^>]+>/g, '') // Remove HTML tags
        .replace(/&nbsp;/g, ' ')
        .replace(/&amp;/g, '&')
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .trim();
      cells.push(cellText);
    }
    
    // Check if this is the totals row
    if (cells.length > 0 && cells[0] === 'Regular Season Stats') {
      totalsRow = cells;
      continue;
    }
    
    // Check if this looks like a game row (has date, opponent, result)
    // Skip header rows
    if (cells.length >= 10 && cells[0] && !cells[0].includes('Date') && !cells[0].includes('Regular Season')) {
      try {
        const game = {
          date: cells[0] || '',
          opponent: cells[1] || '',
          result: cells[2] || '',
          rushing: {
            carries: cells[3] || '0',
            yards: cells[4] || '0',
            avg: cells[5] || '0',
            td: cells[6] || '0'
          },
          receiving: {
            receptions: cells[8] || '0',
            targets: cells[9] || '0',
            yards: cells[10] || '0',
            td: cells[12] || '0'
          }
        };
        games.push(game);
      } catch (e) {
        // Skip malformed rows
        continue;
      }
    }
  }
  
  return { games, totalsRow };
}

/**
 * Scrape ESPN player game log page to extract season statistics
 */
async function scrapeESPNGameLog(playerId, playerName) {
  // Construct URL with slug if player name provided
  let url;
  if (playerName) {
    const slug = playerName.toLowerCase().replace(/ /g, '-').replace(/\./g, '').replace(/'/g, '');
    url = `https://www.espn.com/nfl/player/gamelog/_/id/${playerId}/${slug}`;
  } else {
    url = `https://www.espn.com/nfl/player/gamelog/_/id/${playerId}`;
  }
  
  try {
    const html = await fetchUrl(url);
    const { games, totalsRow } = parseGameLogTable(html);
    
    // Extract player name from page title if not provided
    let extractedName = playerName;
    if (!extractedName) {
      const titleMatch = html.match(/<title[^>]*>(.*?)<\/title>/);
      if (titleMatch) {
        const titleText = titleMatch[1];
        const nameMatch = titleText.match(/^([^-|]+)/);
        if (nameMatch) {
          extractedName = nameMatch[1].replace(' Game Log', '').trim();
        }
      }
    }
    
    // Extract season totals from totals row
    let totalRushCar = 0, totalRushYds = 0, totalRushAvg = 0, totalRushTd = 0;
    
    if (totalsRow && totalsRow.length >= 5) {
      // Format: ['Regular Season Stats', '222', '1,025', '4.6', '10', ...]
      totalRushCar = parseInt((totalsRow[1] || '0').replace(/,/g, '')) || 0;
      totalRushYds = parseInt((totalsRow[2] || '0').replace(/,/g, '')) || 0;
      totalRushAvg = parseFloat((totalsRow[3] || '0')) || 0;
      totalRushTd = parseInt((totalsRow[4] || '0').replace(/,/g, '')) || 0;
    } else if (games.length > 0) {
      // Fallback: calculate from games
      totalRushYds = games.reduce((sum, g) => {
        const yds = parseInt(g.rushing.yards) || 0;
        return sum + yds;
      }, 0);
      totalRushTd = games.reduce((sum, g) => {
        const td = parseInt(g.rushing.td) || 0;
        return sum + td;
      }, 0);
      totalRushCar = games.reduce((sum, g) => {
        const car = parseInt(g.rushing.carries) || 0;
        return sum + car;
      }, 0);
      totalRushAvg = totalRushCar > 0 ? Math.round((totalRushYds / totalRushCar) * 10) / 10 : 0;
    }
    
    // Get most recent game (last in list - games are chronological)
    const lastGame = games.length > 0 ? games[games.length - 1] : null;
    
    // Build result JSON (same format as Python version)
    const result = {
      success: true,
      playerName: extractedName || 'Unknown',
      playerId: playerId,
      seasonStats: {
        rushing: {
          carries: totalRushCar,
          yards: totalRushYds,
          touchdowns: totalRushTd,
          average: totalRushAvg
        }
      },
      lastGame: lastGame ? {
        date: lastGame.date,
        opponent: lastGame.opponent,
        result: lastGame.result,
        rushing: {
          carries: parseInt(lastGame.rushing.carries) || 0,
          yards: parseInt(lastGame.rushing.yards) || 0,
          touchdowns: parseInt(lastGame.rushing.td) || 0,
          average: parseFloat(lastGame.rushing.avg) || 0
        }
      } : null,
      source: 'espn_gamelog',
      gamesPlayed: games.length
    };
    
    return result;
  } catch (error) {
    return {
      success: false,
      error: error.message,
      playerId: playerId
    };
  }
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length < 1) {
    console.log(JSON.stringify({
      success: false,
      error: 'Player ID required as argument'
    }));
    process.exit(1);
  }
  
  const playerId = args[0];
  const playerName = args[1] || null;
  
  const result = await scrapeESPNGameLog(playerId, playerName);
  console.log(JSON.stringify(result));
}

// Run if called directly
if (require.main === module) {
  main().catch(error => {
    console.log(JSON.stringify({
      success: false,
      error: error.message
    }));
    process.exit(1);
  });
}

